library/trans.c

```diff
@@ -557,9 +557,88 @@ int dnet_trans_iterate_move_transaction(struct dnet_net_state *st, struct list_h
 	return trans_moved;
 }
 
+struct dnet_ping_node_private
+{
+	struct dnet_session	*session;
+	struct dnet_net_state	*net_state;
+};
+
+static int dnet_ping_stall_node_complete(struct dnet_addr *addr __unused, struct dnet_cmd *cmd, void *priv)
+{
+	struct dnet_ping_node_private *ping_private;
+	struct dnet_net_state *st;
+
+	if (!is_trans_destroyed(cmd)) {
+	        ping_private = priv;
+		st = ping_private->net_state;
+
+		dnet_session_destroy(ping_private->session);
+		free(ping_private);
+
+		if (cmd->status == -ETIMEDOUT)
+			dnet_state_reset(st, cmd->status);
```

Comment:
If callback is invoked from dnet_session_destroy() called from dnet_ping_stall_node() this will deadlock, since dnet_ping_stall_node() is called under n->state_lock